R/GeospatialFunctions.R

```diff
@@ -203,44 +229,303 @@ fetchATTAINS <- function(.data) {
+        #tibble::rowid_to_column(var = "index")
+      } else {
+        geospatial_data <- .data
+      }
+    } else {
+      # ... Otherwise transform into a spatial object then do the same thing:
+      geospatial_data <- .data %>%
+        # convert dataframe to a spatial object
+        TADA_MakeSpatial(.data = ., crs = 4326) %>%
+        # add unique WQP ID for identifying obs with more than one ATTAINS assessment unit
+        #tibble::rowid_to_column(var = "index") %>%
+        dplyr::mutate(geometry_join = geometry)
+    }
+    
+  }))
+  
+  # Reduce WQP data to unique coordinates 
+  unique_sites <- dplyr::distinct(geospatial_data, geometry)
+  
+  # If user wants HighRes NHD...
+  if(resolution %in% c("NHDPlus HiRes", "NHDPlus HR", "HighRes", "NHDHR", "HR", "NHD HR")){
+    
+    suppressMessages(suppressWarnings({
+      # ... we first must identify the HUC4s that contain the WQP data...   
+      
+      geospatial_aoi <- unique_sites %>%
+        # convert XY sites into super tiny polygons for "AOI"
+        sf::st_buffer(0.0000001) %>%
+        # convert sf object to sfc object (required for downloading from the map server)
+        sf::st_as_sfc(.)
+      
+      nhd_plus_hr_url <- "https://hydro.nationalmap.gov/arcgis/rest/services/NHDPlus_HR/MapServer"
+      
+      # open the nhd_hr - which contains a bunch of layers
+      nhd_hr <- arcgislayers::arc_open(nhd_plus_hr_url)
+      
+      # list the layers of the nhdhr object
+      # list_items(nhd_hr)
+      
+      # select the layer by id from the items list called above
+      nhd_hr_catchments <- arcgislayers::get_layer(nhd_hr, 10)
+      
+      # fields <- list_fields(nhd_hr_waterbody)
+      # View(fields)
+      
+      # you can use SQL-style queries to make this smaller - you have to use the alias
+      # field here.
+      # query <- "Ftype = 390 AND AreaSqKm > 0.1"
+      
+      # use bbox to return associated catchments
+      nhd_catchments <- vector("list", length = length(geospatial_aoi))
+      
+      for(i in 1:length(geospatial_aoi)){
+        try(nhd_catchments[[i]] <- arcgislayers::arc_select(nhd_hr_catchments,
+                                                            # where = query,
+                                                            filter_geom = geospatial_aoi[i],
+                                                            crs = sf::st_crs(geospatial_aoi[i])), silent = TRUE)
+      }
+      
+    }))
+    
+    nhd_catchments <- nhd_catchments %>%
+      purrr::keep(~!is.null(.))
+    
+    try(nhd_catchments <- nhd_catchments %>%
+          dplyr::bind_rows(), silent = TRUE)
+    try(nhd_catchments <- nhd_catchments %>%
+          dplyr::select(nhdplusid,
+                        catchmentareasqkm = areasqkm) %>%
+          dplyr::mutate(nhdplusid = as.character(nhdplusid),
+                        resolution = "HR",
+                        catchmentareasqkm = as.numeric(catchmentareasqkm)) %>%
```

Comment:
consider giving these a preface like: TADA.catchment.nhdplushrid, TADA.catchment.resolution, TADA.catchmentareasqkm or something. Just looking at the resulting dataframe just reminds me it's nice to have the context of where columns originated

Reply:
Now prefaced with "NHD."!